### PR TITLE
Do not exercise diagnostics when disabled

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR allows users to fully disable diagnostics, so that it is not exercised, when disabled. I find myself having to do this locally quite often, so I thought that it might be a good thing to merge.

### CODE REVIEW

**Comment by szy21** [comment]
If the user defines customized diagnostics in the yaml file (e.g. [here](https://github.com/CliMA/ClimaAtmos.jl/blob/3fbb1bccebc8090418eff1f17b0d7f5fc80c77d7/config/model_configs/prognostic_edmfx_bomex_column.yml#L29)) but set `output_default_diagnostics` to false, the model still needs to output diagnostics. We should check that as well?

**Comment by charleskawczynski** [comment]
> If the user defines customized diagnostics in the yaml file (e.g. [here](https://github.com/CliMA/ClimaAtmos.jl/blob/3fbb1bccebc8090418eff1f17b0d7f5fc80c77d7/config/model_configs/prognostic_edmfx_bomex_column.yml#L29)) but set `output_default_diagnostics` to false, the model still needs to output diagnostics. We should check that as well?

There's no documentation/defaults on `diagnostics`. Perhaps we can/should add a separate flag, `enable_diagnostics = true`, which can be used to perform this check. I think it would be convenient to have a single switch

File: src/solver/type_getters.jl [hunk 1]
```diff
@@ -705,13 +709,15 @@ function get_simulation(config::AtmosConfig)
     end
     @info "init integrator: $s"
 
-    s = @timed_str begin
-        integrator = ClimaDiagnostics.IntegratorWithDiagnostics(
-            integrator,
-            scheduled_diagnostics,
-        )
+    if config.parsed_args["output_default_diagnostics"]
```
Comment (Sbozzolo) [inline src/solver/type_getters.jl]:
```suggestion
    if !isempty(scheduled_diagnostics)
```

File: src/solver/type_getters.jl [hunk 2]
```diff
@@ -663,17 +663,21 @@ function get_simulation(config::AtmosConfig)
...
-    s = @timed_str begin
-        scheduled_diagnostics, writers = get_diagnostics(
-            config.parsed_args,
-            atmos,
-            Y,
-            p,
-            t_start,
-            sim_info.dt,
-        )
+    if config.parsed_args["output_default_diagnostics"]
```
Comment (Sbozzolo) [inline src/solver/type_getters.jl]:
```suggestion
if config.parsed_args["output_default_diagnostics"] || !isempty(config.parsed_args["diagnostics"])
```

Comment (Sbozzolo) [reply src/solver/type_getters.jl]: (Should be something like this, I haven't tried)

File: config/default_configs/default_config.yml [hunk 3]
```diff
@@ -259,6 +259,9 @@ override_τ_precip:
 output_default_diagnostics:
   help: "Output the default diagnostics associated to the selected atmospheric model"
   value: true
+enable_diagnostics:
+  help: "Set to false to fully disable the diagnostics"
+  value: true
```
Comment (Sbozzolo) [inline config/default_configs/default_config.yml:264]:
```suggestion
```

File: src/solver/type_getters.jl [hunk 4]
```diff
@@ -663,17 +663,21 @@ function get_simulation(config::AtmosConfig)
...
-    s = @timed_str begin
-        scheduled_diagnostics, writers = get_diagnostics(
-            config.parsed_args,
-            atmos,
-            Y,
-            p,
-            t_start,
-            sim_info.dt,
-        )
+    if config.parsed_args["enable_diagnostics"]
```
Comment (Sbozzolo) [inline src/solver/type_getters.jl:666]:
```suggestion
    if config.parsed_args["enable_diagnostics"] || !isempty(config.parsed_args["diagnostics"]
```

File: src/solver/type_getters.jl [hunk 5]
```diff
@@ -663,17 +663,21 @@ function get_simulation(config::AtmosConfig)
...
+                atmos,
+                Y,
+                p,
+                t_start,
+                sim_info.dt,
+            )
+        end
+        @info "initializing diagnostics: $s"
+    else
+        writers = nothing
```
Comment (Sbozzolo) [inline src/solver/type_getters.jl:679]:
```suggestion
        writers, scheduled_diagnostics = nothing, nothing
```

File: src/solver/type_getters.jl [hunk 6]
```diff
@@ -705,13 +709,15 @@ function get_simulation(config::AtmosConfig)
     end
     @info "init integrator: $s"
 
-    s = @timed_str begin
-        integrator = ClimaDiagnostics.IntegratorWithDiagnostics(
-            integrator,
-            scheduled_diagnostics,
-        )
+    if config.parsed_args["enable_diagnostics"]
```
Comment (Sbozzolo) [inline src/solver/type_getters.jl:712]:
```suggestion
    if !isnothing(scheduled_diagnostics)
```

**Review by Sbozzolo — COMMENTED** [review]
Maybe the more important question is, why do you find yourself disabling this piece of code? What needs to be changed so that you don't have to do that?

**Comment by Sbozzolo** [comment]
> > If the user defines customized diagnostics in the yaml file (e.g. [here](https://github.com/CliMA/ClimaAtmos.jl/blob/3fbb1bccebc8090418eff1f17b0d7f5fc80c77d7/config/model_configs/prognostic_edmfx_bomex_column.yml#L29)) but set `output_default_diagnostics` to false, the model still needs to output diagnostics. We should check that as well?
> 
> There's no documentation/defaults on `diagnostics`. Perhaps we can/should add a separate flag, `enable_diagnostics = true`, which can be used to perform this check. I think it would be convenient to have a single switch

There is documentation. It's here: https://clima.github.io/ClimaAtmos.jl/dev/diagnostics/

And I don't think we should add another redundant switch that effectively just disables other yaml configurations: it makes handling interaction between configurations harder to predict and maintain.

File: src/solver/type_getters.jl (hunk 6 above)
Comment (charleskawczynski) [reply src/solver/type_getters.jl:712]: I don't want to exercise this code, that is the point of this PR.

Comment (Sbozzolo) [reply src/solver/type_getters.jl:712]: I don't understand. If you add the suggestion above, this code is not exercised when there are no diagnostics

Comment (Sbozzolo) [reply src/solver/type_getters.jl:712]: So, if you want to not exercise this code, you can just comment out the diagnostics from your configuration

**Comment by charleskawczynski** [comment]
> And I don't think we should add another redundant switch that effectively just disables other yaml configurations: it makes handling interaction between configurations harder to predict and maintain.

It's not redundant in that it ensures that our diagnostics are not exercised, the other flags do not do this. Think of this as a developer flag

**Comment by Sbozzolo** [comment]
> > Maybe the more important question is, why do you find yourself disabling this piece of code?
> 
> Because the diagnostics is still brittle. I already gave feedback on the diagnostics and `IntegratorWithDiagnostics` design, and you insisted to keep it. All I'm asking for is a way to disable it now, because it's frequently causing problems for me during development. IIRC, the last error I saw was an assumption about the orography type. @bloops recently ran into an issue when we were trying to run ClimaAtmos on the cloud. I'm happy to open issues when they arise, but these snags are slowing development for me and others.

I am happy to have this PR merged, but please, report these issues and tag me so that we can fix them. I didn't know you or @bloops had problems.

> It's not redundant in that it ensures that our diagnostics are not exercised, the other flags do not do this. Think of this as a developer flag

I think that my suggestion is equivalent in terms of code not being exercised but does not add an extra configuration that needs to be maintained.

**Review by Sbozzolo — APPROVED** [review]
> Thats fair, but it does require users to specify both flags to disable diagnostics (and the `diagnostics` flag does not seem to have a default), and those may or may not live in two different files for a particular configuration. I'm happy to (eventually) remove this flag, but I don't see it as adding major overhead, and I think that the name and docs are clear.

It's not a major overhead, but I would like to try resisting configuration creep, especially when other flags are very close in scope. We have already more than 100 and it is so easy to end up with configurations that are misconfigured and find unexpected results.


The default value for `diagnostics` is its absence. `diagnostics` is not a flag, but a nested structure, and it cannot be described in the same way we use for the other flags in `default_config.yml`. We can probably add it by having it parse as `nothing` by default, and handle that case specially.

**Pull request merged**